thirdparty/patches/jemalloc_nodump.patch

```diff
@@ -0,0 +1,222 @@
+ CTL_PROTO(arena_i_dirty_decay_ms)
+@@ -494,6 +495,7 @@ static const ctl_named_node_t arena_i_node[] = {
+ 	{NAME("purge"),		CTL(arena_i_purge)},
+ 	{NAME("reset"),		CTL(arena_i_reset)},
+ 	{NAME("destroy"),	CTL(arena_i_destroy)},
++	{NAME("dontdump"),	CTL(arena_i_dontdump)},
+ 	{NAME("dss"),		CTL(arena_i_dss)},
+ 	/*
+ 	 * Undocumented for now, since we anticipate an arena API in flux after
+@@ -2678,6 +2680,67 @@ arena_i_reset_ctl(tsd_t *tsd, const size_t *mib, size_t miblen, void *oldp,
+ 	return ret;
+ }
+
++static void
++arena_i_dontdump(tsdn_t *tsdn, unsigned arena_ind) {
++    malloc_mutex_lock(tsdn, &ctl_mtx);
++    {
++        unsigned narenas = ctl_arenas->narenas;
++        /*
++	 * Access via index narenas is deprecated, and scheduled for
++	 * removal in 6.0.0.
++         */
++        if (arena_ind == MALLCTL_ARENAS_ALL || arena_ind == narenas) {
++            unsigned i;
++            VARIABLE_ARRAY(arena_t *, tarenas, narenas);
++            for (i = 0; i < narenas; i++) {
++                tarenas[i] = arena_get(tsdn, i, false);
++            }
++
++            /*
++             * No further need to hold ctl_mtx, since narenas and
++             * tarenas contain everything needed below.
++             */
++            malloc_mutex_unlock(tsdn, &ctl_mtx);
++
++            for (i = 0; i < narenas; i++) {
++                if (tarenas[i] != NULL) {
++                    arena_dontdump(tsdn, tarenas[i]);
++                }
++            }
++        } else {
++            arena_t *tarena;
++
++            assert(arena_ind < narenas);
++
++            tarena = arena_get(tsdn, arena_ind, false);
++
++            /* No further need to hold ctl_mtx. */
++            malloc_mutex_unlock(tsdn, &ctl_mtx);
++
++            if (tarena != NULL) {
++                arena_dontdump(tsdn, tarena);
++            }
++        }
++    }
++
++}
++
++static int
++arena_i_dontdump_ctl(tsd_t *tsd, const size_t *mib, size_t miblen, void *oldp,
++    size_t *oldlenp, void *newp, size_t newlen) {
++
++        int ret = 0;
++        unsigned arena_ind;
++        NEITHER_READ_NOR_WRITE();
++        MIB_UNSIGNED(arena_ind, 1);
++
++        arena_i_dontdump(tsd_tsdn(tsd), arena_ind);
++
++label_return:
++        return ret;
++}
++
++
+ static int
+ arena_i_destroy_ctl(tsd_t *tsd, const size_t *mib, size_t miblen, void *oldp,
+     size_t *oldlenp, void *newp, size_t newlen) {
+diff --git src/extent.c src/extent.c
+index cf3d1f31..adf0d574 100644
+--- src/extent.c
++++ src/extent.c
+@@ -147,6 +147,21 @@ ecache_dalloc(tsdn_t *tsdn, pac_t *pac, ehooks_t *ehooks, ecache_t *ecache,
+ 	extent_record(tsdn, pac, ehooks, ecache, edata);
+ }
+
++void
++ecache_dontdump(tsdn_t *tsdn, ecache_t *ecache) {
++    malloc_mutex_lock(tsdn, &ecache->mtx);
++
++    eset_t *eset = &ecache->eset;
++    edata_t *edata = edata_list_inactive_first(&eset->lru);
++
++    while (edata != NULL) {
++        pages_dontdump(edata_base_get(edata), edata_size_get(edata));
```

Comment:
Is this function provided by jemalloc already?

Reply:
yes
```
bool
pages_dontdump(void *addr, size_t size) {
	assert(PAGE_ADDR2BASE(addr) == addr);
	assert(PAGE_CEILING(size) == size);
#if defined(JEMALLOC_MADVISE_DONTDUMP)
	return madvise(addr, size, MADV_DONTDUMP) != 0;
#elif defined(JEMALLOC_MADVISE_NOCORE)
	return madvise(addr, size, MADV_NOCORE) != 0;
#else
	return false;
#endif
}
```